specs/governance/alligator.md

```diff
@@ -0,0 +1,390 @@
+from its [storage](#storage). If either address has not been migrated, the `Alligator` MUST copy the delegation
+and checkpoint data from the token contract to its own state. After copying the data, the `Alligator` MUST update
+the `migrated` mapping to reflect that the address has been migrated.
+
+The `afterTokenTransfer` function MUST update the voting power of the `_from` and `_to` addresses with the same logic
+of the `GovernanceToken`.
+
+#### `migrateAccount`
+
+Migrates the delegation state of a given account from the `GovernanceToken` to the `Alligator` contract. This function
+MUST only migrate the `_account` if it has not been migrated yet.
+
+```solidity
+function migrateAccount(address _account) external
+```
+
+### Getters
+
+For backwards compatibility, the `Alligator` MUST implement all public getter functions of the
+`GovernanceToken` related to delegation and voting power. These functions MUST be used by the
+`GovernanceToken` when an account has been been migrated to the `Alligator` contract. Otherwise,
+the `GovernanceToken` MUST use its own state. Similarly, all of the `Alligator` getter functions
+MUST use the `GovernanceToken` state if the account has not been migrated.
+
+#### `checkpoints`
+
+Retrieves the checkpoints for a given user address.
+
+```solidity
+function checkpoints(address _account) external view returns (ERC20Votes.Checkpoint[] memory)
+```
+
+#### `numCheckpoints`
+
+Retrieves the number checkpoints for a given user address.
+
+```solidity
+function numCheckpoints(address _account) external view returns (uint32)
+```
+
+#### `getVotes`
+
+Retrieves the voting power of a given user address.
+
+```solidity
+function getVotes(address _account) external view returns (uint256)
+```
+
+#### `getPastVotes`
+
+Retrieves the voting power of a given user address at a block.
+
+```solidity
+function getPastVotes(address _account, uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `getPastTotalSupply`
+
+Retrieves the total supply of the `GovernanceToken` at a given block.
+
+```solidity
+function getPastTotalSupply(uint256 _blockNumber) external view returns (uint256)
+```
+
+#### `subdelegations`
+
+Retrieves the subdelegations for a given user and delegatee.
+
+```solidity
+function subdelegations(address _account, address _delegatee) external view returns (SubdelegationRules memory)
+```
+
+### Events
+
+#### `Subdelegation`
+
+MUST trigger when an account subdelegates voting power to another address (delegatee).
+
+```solidity
+event Subdelegation(address indexed account, address indexed delegatee, SubdelegationRules rule)
+```
+
+#### `Subdelegations`
+
+MUST trigger when an account subdelegates voting power to multiple addresses (delegatees).
+
+```solidity
+event Subdelegations(address indexed account, address[] delegatee, SubdelegationRules[] rules)
+```
+
+#### `DelegateVotesChanged`
```

Comment:
This should be the one i've been talking about. The current version does not incorporate changes from the existing alligator contract

Reply:
@chuxinh do you prefer to have two events when a subdelegation is applied (so the `Subdelegation` and `DelegateVotesChanged` emitted at the same time) or only one event (for example `Subdelegation`, which includes the same parameters as `DelegateVotesChanged` -- new balance and old balance). in this second case, the `DelegateVotesChanged` would only be triggered for token transfers and not when a subdelegation rule is applied/updated.

Reply:
updated for the second approach: https://github.com/ethereum-optimism/specs/pull/247/commits/88c104f1a1ce5af225d224cbe324c0f63bcd8a5e

Reply:
Thanks for the clarification here! Let me know if my understanding here is correct
`DelegateVotesChanged` which is the current event emitted in the OP token contract will continue emitting in the new interop OP standard when there's token transfer. And in the current spec, `Subdelegation` will only be emitted when it's a subdelegation but not a token transfer. So in order to figure out a delegate's actual voting power, it needs to be an add up from both events?

Reply:
Also i see that there are two events, one called `Subdelegation` and the other is `Subdelegations`, given that `Subdelegations` seems to emit for batch subdelegation, what's it representing for before and after balance? 

Also wanna make sure event naming are clear enough with no duplications because previously for Alligator V5, third party providers such as Dune had to create custom decoding due to poor event naming. 
There are two events named `subdelegations` and both of these events have a field named `subdelegationsRules`

Reply:
`DelegateVotesChanged` will be emitted every time the voting power of a user changes -- this includes for both token transfers and subdelegation updates.

so, for a token transfer, it will only emit `DelegateVotesChanged` because the subdelegations per se haven't changed.

when a subdelegation is created/updated/removed, it will event the `Subdelegation` event and `DelegateVotesChanged` for every voting power change. this allows to always have visibility of voting power changes, for both token transfers and subdelegation updates.

Reply:
That makes sense. I think it's okay to keep it for now. so for someone who does not care about if the change is coming from subdelegation or not, they can continue using `DelegateVotesChanged`.

Also curious how it's treating the historical records since at the moment, we cannot technically derive the actual voting power without computing the subdelegation rule. How does it work to derive the voting power balance when this goes live?